docs/adr/adr_database-history.md

```diff
@@ -0,0 +1,150 @@
+- dealing with data type conversion when storing/reading change payload ("old"/"new" value)
+
+### History tables (or: shadow tables)
+
+For every relevant table create a history table with identical schema and an additional `effective` date column. On every change in a tracked table, a copy of the updated row is inserted as new record into the history table.
+
+*Advantages*
+
+- change date and user are recorded
+- simple, especially querying (select by highest value of `effective` date column)
+- separation of data and history
+- can be combined with database triggers to record any change on DB level (no change in application code necessary then)
+- history tables should be insert-only; they cannot be modified then
+
+*Drawbacks*
+
+- extensibility costs: every change in the original table requires an update in the corresponding history table
+- the state of foreign key fields (i.e. tables that a history table depends on) has to be recorded as well to still be correct later
+
+### Insert-only tables
+
+Whenever a row would change, it is marked as inactive and copied into a new, active row.
+
+*Advantages*
+
+- no additional tables needed
+
+*Drawbacks*
+
+- no separation of responsibilities (relevant data vs. history)
+- it's not possible to refer to such table entries using foreign keys (no referential integrity)
+
+### Snapshots
+
+Make copies of the database state at repeated times and archive them.
+
+*Advantages*
+
+- copying is simple
+- no extension of business logic required
+
+*Drawbacks*
+
+- analysing is complicated: DB schema might have changed over time
+- storage needs to be defined: in DB or as separate file
+- precise changes are not recorded
+
+### Event sourcing
+
+Every change to the state of an application is captured in an event object. These event objects are themselves stored in the sequence they were applied for the same lifetime as the application state itself. The event log can be used to reconstruct past states.
+
+*Advantages*
+
+- powerful, scalable concept
+- helpful for auditing and debugging (replaying events in a test environment)
+
+*Drawbacks*
+
+- complex extension of codebase required (twice for dropapp and v2 back-end)
+- new developers might need to familiarize themselves with the concept first
+
+## Decision
+
+There are several popular audit logging techniques, but none of them serve every purpose. The most effective ones are often expensive, resource intensive, or performance degrading. Others are cheaper in terms of resources but are either incomplete, cumbersome to maintain, or require a sacrifice in design quality.
+
+I propose to use history tables. It meets the requirements for simplicity and for knowing historical state. Since we don't have a lot of tables to be shadowed (stock, products, locations, people, transactions), the maintenance cost is low.
```

Comment:
I do not think we have to shadow transactions. It is already a "transactional" db table and not a CRUD table. We can figure out the distributed items in the free shop just by counting/summing the rows between two dates. 
One can delete transactions though.

Reply:
I went through the db. In the best case we should implement shadow tables for all db tables that the user can edit and especially the CRUD ones.
This means:

- locations
- people
- products
- qr (since it is INSERT only, we do not need to shadow it, but a created_by column would be nice here)
- stock 
- tags (this is the only N-to-N relationship, which probably means we have to shadow tags_relations, too)
- transactions (since it is INSERT DELETE only, we do not need to shadow it)

not sure what to do about:
- distro_events
- transfer_agreements
- shipments

Stuff we do not want to mirror:
- anything connected to borrow_, laundry_, library_ because we discontinue support for these features.
- Anything user related.